### ISSUE TITLE
Avoid "replacing" original "template" with a "new page"

### ISSUE THREAD

**Issue by ashmaroli** [sverrirs/jekyll-paginate-v2#134]
This ticket is mainly to **collect your thoughts behind** a fundamental route in this plugin..
https://github.com/sverrirs/jekyll-paginate-v2/blob/b7cd20c805fa01b681b50516dbb676608cc25538/lib/jekyll-paginate-v2/generator/paginationModel.rb#L277-L278

..and to consecutively implement a new route that incorporates your requirements without destroying the original object.

## Why this is an "issue" for me
I have a plugin that works based on special attributes (defined at runtime) of a subclassed instance of `Jekyll::Page` / `Jekyll::Document`.
When this plugin replaces the "original template object", the "special attributes" of the "original template object" are lost and therefore creating conflicts with my plugin.

## Objective
Once I get to know your thoughts behind the current design, I can opt for one of the following:
  - If your design requirements "can be achieved" without "replacing the original template object", then I'll attempt to open a PR to resolve this issue with minimal incompatibilities.
  - If your design requirements "cannot be achieved" by an "alternate approach", then I can simply document this plugin as *`"known to be incompatible"`* with my plugin.

**Comment by sverrirs** [comment]
I'm not quite sure what kind of changes you are suggesting but I'm very much in favour of increasing compatibility with other plugins. :) 

However, whatever route we take I'd like this plugin to not have a hard dependency on other Jekyll plugins or incorporate special code paths or code around idiosyncratic behavior of other plugins.

I think Boris #105 came up with a solution to propagate settings into the pagination pages that may have some merit due to its simplicity.

The over all design I was hoping to achieve was that the Pagination would be the absolute last thing that runs and therefore all other plugins had already set their front matter etc and the pagination could safely just propigate what is already set.

**Comment by ashmaroli** [comment]
> I'm not quite sure what kind of changes you are suggesting

Sorry. I'll try to elaborate further.
Say I have a site, with just a single `Jekyll::Page` instance pointing to path `web/index.md`.
Now, under Vanilla Jekyll:
```
site.pages = [
  #<Jekyll:Page @name="index.md"
]
```

And with this plugin paginating the page into two pages:
```
site.pages = [
  #<Jekyll::PaginateV2::Generator::PaginationPage @name="index.html">,
  #<Jekyll::PaginateV2::Generator::PaginationPage @name=".html">
]
```

What I'd like to get:

under Vanilla Jekyll:
```
site.pages = [
  #<Jekyll:Page @name="index.md"
]
```

And with this plugin paginating the page into two pages:
```
site.pages = [
  #<Jekyll:Page @name="index.md",
  #<Jekyll::PaginateV2::Generator::PaginationPage @name=".html">
]
```

### &nbsp;

> I'd like this plugin to not have a hard dependency on other Jekyll plugins or incorporate special code paths..

Agreed. I was not hinting that it should..

>  or code around idiosyncratic behavior of other plugins.

Agreed. That would be very impractical.

> came up with a solution to propagate settings into the pagination pages

That solution though is entirely unrelated to this issue.

> hoping to achieve was that the Pagination would be the absolute last thing that runs

Yes, that was implied by the generator's `:prority`

> the pagination could safely just propigate what is already set.

Unfortunately, attributes are not entirely getting transferred. Well, that can be *ignored* within the context of this ticket..

## &nbsp;

So, getting straight to the point, I'd like answers to the following:
- What "special attributes" do the `first_index_page` **need to have** for this plugin to work as expected..?
- Would you be open to simply having the "original template" `respond_to?` the above "special attributes" in addition to its existing attributes..? (The onus of implementing the idea is on me.)

**Comment by sverrirs** [comment]
The first page is no different than the N others if I remember correctly (let me double check this though). It needs the same attributes and the pagination variables as the others. (this is why I discarded the original one as it was just there as a yeah "template" containing the user's config. There were some practical issues that blocked me simply reusing the original "template" I think primarily the page inheritance that I added to the pagination Pages.

There is nothing really from me stopping us from restructuring this. However I'd like the restructuring effort to be a "larger picture" effort, i.e. the pagination plugin currently works well for a broad range of solutions. If your plugin is the only one that doesn't work with it I'd like to investigate also if the problem could be resolved easier with changes in yours (or perhaps a little bit on both sides) :)
